src/main/java/com/iota/iri/network/NeighborRouter.java

```diff
@@ -0,0 +1,840 @@
+     * IP address is added through {@link NeighborRouter#addNeighbor(String)}.
+     * 
+     * @param ipAddress       The IP address
+     * @param newNeighborConn The {@link SocketChannel} to close if the connection is not allowed
+     * @return true if allowed, false if not
+     * @throws IOException if anything goes wrong closing the {@link SocketChannel}
+     */
+    private boolean okToConnect(String ipAddress, SocketChannel newNeighborConn) throws IOException {
+        if (hostsBlacklist.contains(ipAddress)) {
+            // silently drop connection
+            newNeighborConn.close();
+            return false;
+        }
+        if (availableNeighborSlotsFilled()) {
+            log.info("dropping new connection from {} as all neighbor slots are filled", ipAddress);
+            newNeighborConn.close();
+            return false;
+        }
+        boolean whitelisted = hostsWhitelist.contains(ipAddress);
+        if (!whitelisted) {
+            if (!config.isAutoTetheringEnabled()) {
+                log.info("blacklisting/dropping new connection as neighbor from {} is not defined in the config",
+                        ipAddress);
+                hostsBlacklist.add(ipAddress);
+                newNeighborConn.close();
+                return false;
+            }
+            log.info("new auto-tethering connection from {}", ipAddress);
+        }
+        return true;
+    }
+
+    /**
+     * Closes the connection to the neighbor, re-registers the {@link ServerSocketChannel} for
+     * {@link SelectionKey#OP_CONNECT} in case neighbor slots will be available again and finally removes the neighbor
+     * from the connected neighbors map.
+     * 
+     * @param channel  {@link SocketChannel} to close
+     * @param identity The identity of the neighbor, null must be passed if the neighbor should not be marked as not
+     *                 connected.
+     * @param selector The used {@link Selector}
+     */
+    private void closeNeighborConnection(SelectableChannel channel, String identity, Selector selector) {
+        try {
+            channel.close();
+        } catch (IOException e) {
+            return;
+        }
+        // re-register the server socket for incoming connections as we will have a new slot open
+        if (availableNeighborSlotsFilled()) {
+            SelectionKey key = serverSocketChannel.keyFor(selector);
+            if (key != null) {
+                key.interestOps(SelectionKey.OP_ACCEPT);
+            }
+        }
+        if (identity == null) {
+            return;
+        }
+        if (connectedNeighbors.remove(identity) != null) {
+            log.info("removed neighbor {} from connected neighbors", identity);
+        }
+    }
+
+    private boolean availableNeighborSlotsFilled() {
+        // while this check if not thread-safe, initiated connections will be dropped
+        // when their handshaking was done but already all neighbor slots are filled
```

Comment:
This check is also performed on `addNeighbor` which can be called from any thread, hence, this operation is not thread-safe. However, they way the NeighborRouter is written, there will never be the case where the node has more "connected neighbors" than defined in the config, since addition and removal from this map are from within the single-threaded selector loop.